app/assets/javascripts/discourse/tests/integration/components/async-content-test.gjs

```diff
@@ -0,0 +1,350 @@
+      const promise = Promise.resolve("data returned");
+
+      await render(<template>
+        <AsyncContent @asyncData={{promise}}>
+          <:content as |data|>
+            <div class="content">
+              {{data}}
+            </div>
+          </:content>
+        </AsyncContent>
+      </template>);
+
+      assert.dom(".content").exists();
+      assert.dom(".content").hasText("data returned");
+    });
+
+    test("it does not display the block if the promise fails", async function (assert) {
+      const promise = Promise.reject("error");
+
+      await render(<template>
+        <AsyncContent @asyncData={{promise}}>
+          <:content as |data|>
+            <div class="content">
+              {{data}}
+            </div>
+          </:content>
+        </AsyncContent>
+        <DialogHolder />
+      </template>);
+
+      assert.dom(".content").doesNotExist();
+    });
+  });
+
+  module("<:empty>", function () {
+    test("it displays the block when the promise is resolved with an empty value", async function (assert) {
+      const promise = Promise.resolve(null);
+
+      await render(<template>
+        <AsyncContent @asyncData={{promise}}>
+          <:empty>
+            <div class="empty">
+              Empty
+            </div>
+          </:empty>
+        </AsyncContent>
+      </template>);
+
+      assert.dom(".empty").exists();
+    });
+
+    test("it does not display the block when the promise is resolved with a value", async function (assert) {
+      const promise = Promise.resolve("data");
+
+      await render(<template>
+        <AsyncContent @asyncData={{promise}}>
+          <:empty>
+            <div class="empty">
+              Empty
+            </div>
+          </:empty>
+        </AsyncContent>
+      </template>);
+
+      assert.dom(".empty").doesNotExist();
+    });
+
+    test("it displays the content block if the the empty block is not provided", async function (assert) {
+      const promise = Promise.resolve(null);
+
+      await render(<template>
+        <AsyncContent @asyncData={{promise}}>
+          <:content>
+            <div class="content">
+              Empty
+            </div>
+          </:content>
+        </AsyncContent>
+      </template>);
+
+      assert.dom(".content").exists();
+    });
+
+    test("it does not display the block if the promise fails", async function (assert) {
+      const promise = Promise.reject("error");
+
+      await render(<template>
+        <AsyncContent @asyncData={{promise}}>
+          <:empty>
+            <div class="empty">
+              Empty
+            </div>
+          </:empty>
+        </AsyncContent>
+        <DialogHolder />
+      </template>);
+
+      assert.dom(".empty").doesNotExist();
+    });
+  });
+
+  module("<:error>", function () {
+    test("it displayes a popup error dialog when the block is not provided", async function (assert) {
```

Comment:
I wonder whether we should be showing the message inline rather than popping a dialog? I think that would make more sense for most things. What do you think?

IIRC the 'parsing' logic from popupAjaxError is available in a separate function.

Reply:
We have `@errorModes` now. 😃